sklearn/tree/tests/test_tree.py

```diff
@@ -2524,20 +2563,31 @@ def make_friedman1_classification(*args, **kwargs):
 
 
 @pytest.mark.parametrize(
-    "make_data,Tree",
+    "make_data, Tree, tolerance",
     [
-        (datasets.make_friedman1, DecisionTreeRegressor),
-        (make_friedman1_classification, DecisionTreeClassifier),
+        # Due to the sine link between X and y, we expect the native handling of
+        # missing values to always be better than the naive mean imputation in the
+        # regression case.
+        #
+        # Due to randomness in ExtraTree, we expect the native handling of missing
+        # values to be sometimes better than the naive mean imputation, but not always
+        (datasets.make_friedman1, DecisionTreeRegressor, 0),
+        (datasets.make_friedman1, ExtraTreeRegressor, 0),
+        (make_friedman1_classification, DecisionTreeClassifier, 0.02),
+        (make_friedman1_classification, ExtraTreeClassifier, 0.05),
     ],
 )
 @pytest.mark.parametrize("sample_weight_train", [None, "ones"])
 def test_missing_values_is_resilience(
-    make_data, Tree, sample_weight_train, global_random_seed
+    make_data, Tree, sample_weight_train, global_random_seed, tolerance
 ):
     """Check that trees can deal with missing values have decent performance."""
```

Comment:
It seems like this tests fails now because we have removed the extra branch which tries to get a better split with sending all missing values to the right. Now, since we are choosing whether to send missing values to the right or left randomly, sometimes if the split is not that good the overall score might also decrease and using global_random_seed tests for a variety of different random states. 

Either we might need to remove global_random_seed from this test or allow for a more flexible threshold. I tried out using max_depth = 20 here but that still results in 10 tests failing on my local system all related to ExtraTreeClassifier. 

For one test the threshold that needs to be specified is around 0.09
```console
Tree = <class 'sklearn.tree._classes.ExtraTreeClassifier'>
sample_weight_train = None, global_random_seed = 33, tolerance = 0.05
AssertionError: score_native_tree=0.664 + 0.05 should be strictly greater than 0.7536
```
Even then it seems like sometimes there can be an unexpectedly worse result on some reed on a new run.

Reply:
I expect that `ExtraTree*` is only used in the context of a forest.

For any single `ExtraTree*`, I do not expect it to perform well even if a regular non-missing feature is informative (because the threshold is picked at random). By extension, if missing-ness is informative then it should also not perform well with single `ExtraTree*`.

I'm +0.5 in favor of keep the extra branch.

Reply:
I'll let @OmarManzoor and @glemaitre weigh in real quick. 

Should I:

1. Have an if/else code block where we compute criterion again for all missing values down one side and all non missing values down another branch. Cons: one extra criterion evaluation for features with missing values. Pros: performance better. 

2. Not have the if/else code block and only sample random threshold on non missing value range. Send missing values randomly to left or right side. Cons: doesn't account for this extra branching possibility. Pros: same computation as before on main.

Reply:
Okay for now, I reverted the change back to evaluate this extra if/else branch, since @thomasjpfan and I are both +0.5.

[90e8cec](https://github.com/scikit-learn/scikit-learn/pull/27966/commits/90e8cec8ea886fbd62ae28aa4a2e486a3447dd81)

Reply:
Ah I mistyped. I intended the opposite: +0.5 on not having the additional branch.

My argument for removing the branch is that it fits better with the Extra Tree algorithm. By design, it should be random when splitting.

Reply:
Ah I see. Okay, I'm okay with that and tested with SKLEARN_GLOBAL_RANDOM_SEED and there is still predictive power, but obviously more noisy and variance. This problem goes away with the forests, so I'm okay with that.

I've subsequently updated the unit-tests to reflect this removal and also the documentation of the trees.

[96972c0](https://github.com/scikit-learn/scikit-learn/pull/27966/commits/96972c0500719ff4127fdc94d3811167d654182c)

Reply:
I think keeping the randomness seems suitable 👍 , but we should just be mindful of the fact that because of the randomness sometimes it won't produce better results.